ddtrace/tracer/stats.go

```diff
@@ -0,0 +1,324 @@
+}
+
+func (s *rawGroupedStats) export(k aggregation) (groupedStats, error) {
+	msg := s.okDistribution.ToProto()
+	okSummary, err := proto.Marshal(msg)
+	if err != nil {
+		return groupedStats{}, err
+	}
+	msg = s.errDistribution.ToProto()
+	errSummary, err := proto.Marshal(msg)
+	if err != nil {
+		return groupedStats{}, err
+	}
+	// round a float to an int, uniformly choosing
+	// between the lower and upper approximations.
+	round := func(f float64) uint64 {
+		i := uint64(f)
+		if rand.Float64() < f-float64(i) {
+			i++
+		}
+		return i
+	}
+	return groupedStats{
+		Service:        k.Service,
+		Name:           k.Name,
+		Resource:       k.Resource,
+		HTTPStatusCode: k.StatusCode,
+		Type:           k.Type,
+		Hits:           round(s.hits),
+		Errors:         round(s.errors),
+		Duration:       round(s.duration),
+		TopLevelHits:   round(s.topLevelHits),
+		OkSummary:      okSummary,
+		ErrorSummary:   errSummary,
+		Synthetics:     k.Synthetics,
+	}, nil
+}
+
+func newRawBucket(btime uint64) *rawBucket {
+	return &rawBucket{
+		start:    btime,
+		duration: uint64(bucketSize),
+		data:     make(map[aggregation]*rawGroupedStats),
+	}
+}
+
+func (sb *rawBucket) handleSpan(ss *spanSummary) {
+	key := aggregation{
+		Name:       ss.Name,
+		Env:        ss.Env,
+		Type:       ss.Type,
+		Resource:   ss.Resource,
+		Service:    ss.Service,
+		Hostname:   ss.Hostname,
+		StatusCode: ss.StatusCode,
+		Version:    ss.Version,
+		Synthetics: ss.Synthetics,
+	}
+	gs, ok := sb.data[key]
+	if !ok {
+		gs = newRawGroupedStats()
+		sb.data[key] = gs
+	}
+	if ss.TopLevel {
+		gs.topLevelHits += 1
+	}
+	gs.hits += 1
+	if ss.Error != 0 {
+		gs.errors += 1
+	}
+	gs.duration += float64(ss.Duration)
+	// alter resolution of duration distro
+	trundur := nsTimestampToFloat(ss.Duration)
+	if ss.Error != 0 {
+		gs.errDistribution.Add(trundur)
+	} else {
+		gs.okDistribution.Add(trundur)
+	}
+}
+
+// Export transforms a RawBucket into a statsBucket, typically used
+// before communicating data to the API, as RawBucket is the internal
+// type while statsBucket is the public, shared one.
+func (sb *rawBucket) Export() statsBucket {
+	csb := statsBucket{
+		Start:    sb.start,
+		Duration: sb.duration,
+		Stats:    make([]groupedStats, len(sb.data)),
+	}
+	for k, v := range sb.data {
+		b, err := v.export(k)
+		if err != nil {
+			log.Error("Could not export stats bucket: %v.", err)
+			continue
+		}
+		csb.Stats = append(csb.Stats, b)
+	}
+	return csb
+}
+
+// nsTimestampToFloat converts a nanosec timestamp into a float nanosecond timestamp truncated to a fixed precision
+func nsTimestampToFloat(ns int64) float64 {
+	// 10 bits precision (any value will be +/- 1/1024)
+	const roundMask int64 = 1 << 10
+	var shift uint
+	for ns > roundMask {
+		ns = ns >> 1
+		shift++
+	}
+	return float64(ns << shift)
+}
+
+const (
+	// relativeAccuracy is the value accuracy we have on the percentiles. For example, we can
+	// say that p99 is 100ms +- 1ms
+	relativeAccuracy = 0.01
```

Comment:
We can align relativeAccuracy with what we use in the backend store to guarantee it. We need to make the same change in the agent. In the backend it's `defaultEps      = 1.0 / 128.0`

Reply:
Do you mean `relativeAccuracy = 1.0 / 128.0` ? The `defaultEps` part confused me.

Reply:
@piochelepiotr ?